File: Datasets/TrickyBugs/GenProgs/dpp_generated_progs_cpp/p03846/p03846_num0_parsed.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#define ll long long
#define MOD 1000000007
using namespace std;

int main() {
    int N;
    cin >> N;
    vector<int> A(N);
    for (int i = 0; i < N; i++) {
        cin >> A[i];
    }
    
    int cnt_odd = 0, cnt_even = 0;
    for (int i = 0; i < N; i++) {
        if (i % 2 == 0) {
            cnt_odd += A[i] / 2;
            cnt_even += (A[i] + 1) / 2;
        } else {
            cnt_odd += (A[i] + 1) / 2;
            cnt_even += A[i] / 2;
        }
    }
    
    if (cnt_odd != cnt_even) {
        cout << 0 << endl;
        return 0;
    }
    
    vector<ll> factorial(N + 1);
    factorial[0] = 1;
    for (int i = 1; i <= N; i++) {
        factorial[i] = (factorial[i - 1] * i) % MOD;
    }
    
    ll res = 1;
    for (int i = 0; i < N; i++) {
        res = (res * factorial[A[i]]) % MOD;
    }
    
    ll ans = (factorial[N] * res) % MOD;
    cout << ans << endl;
    
    return 0;
}

File: Datasets/TrickyBugs/GenProgs/dpp_generated_progs_python/p03846/p03846_num0_parsed.py
from math import factorial

MOD = 10**9+7

def pow_mod(a, b):
    res = 1
    while b > 0:
        if b & 1:
            res = (res * a) % MOD
        a = (a * a) % MOD
        b >>= 1
    return res

def inverse_mod(a):
    return pow_mod(a, MOD-2)

def calculate_possible_orders(N, A):
    max_diff = N - 1
    freq = [0] * (max_diff + 1)

    for i in range(N):
        if A[i] > max_diff or freq[A[i]] > 0:
            return 0
        freq[A[i]] += 1

    ans = factorial(N)
    for i in range(1, max_diff + 1):
        ans = (ans * inverse_mod(pow_mod(i, freq[i]))) % MOD

    return ans

# Read input
N = int(input())
A = list(map(int, input().split()))

# Calculate and print the number of possible orders
print(calculate_possible_orders(N, A))